Per nodegroup scale-down config

MaciekPytel opened the pull request:
This PR implements https://github.com/kubernetes/autoscaler/issues/3583#issuecomment-743215343. It allows NodeGroups to override ScaleDownUnneededTime, ScaleDownUnreadyTime and utilization thresholds for nodes in that NodeGroup.

A new GetOptions method is added to cloudprovider interface. Its implementation is optional - returning ErrNotImplemented will result in the current behavior of using global config for all NodeGroups. A stub implementation returning ErrNotImplemented was added to all providers.

Note that this PR is only a framework for implementing the actual feature - there is no implementation for how the user would specify the options. Any provider is free to implement GetOptions (and the way to configure it) as they like.

MaciekPytel commented:
/hold
I want to add some more unittests.

MaciekPytel commented:
/hold cancel

BigDarkClown commented on cluster-autoscaler/processors/nodegroupconfig/node_group_config_processor_test.go:
```diff
@@ -0,0 +1,194 @@
+				testUnreadyTime(t, p, c, ng, w, we)
+				testUtilizationThreshold(t, p, c, ng, w, we)
+				testUnneededTime(t, p, c, ng, w, we)
+				testUnneededTime(t, p, c, ng, w, we)
```
These two additional `testUnneededTime` executions seem to be unnecessary.

MaciekPytel replied:
I added them to test that multiple executions of the same function result in the same result and that calling other functions in the meantime doesn't impact it. Would moving it to separate test case make it more readable?

BigDarkClown replied:
I think that would help, right now it seems more like a copy+paste error than a tested case.

MaciekPytel replied:
moved to separate case

Kafei59 commented on cluster-autoscaler/config/autoscaling_options.go:
```diff
@@ -30,10 +30,9 @@ type GpuLimits struct {
-	MaxEmptyBulkDelete int
+// NodeGroupAutoscalingOptions contain various options to customize how autoscaling of
+// a given NodeGroup works. Different options can be used for each NodeGroup.
+type NodeGroupAutoscalingOptions struct {
```
I may have a dumb question but:
- Why only these 4 scale down parameters ?
`ScaleDownDelayAfterAdd`, `ScaleDownDelayAfterDelete` and so on, are not applicable for this new configuration method ?

MaciekPytel replied:
Those 4 values were the low-hanging fruits - they required relatively small changes in the code and they were by far the most commonly requested values to control per nodegroup.

I think ultimately it makes sense to have ScaleDownDelayAfter* as part of NodeGroupAutoscalingOptions, but it is much more tricky to do (details below) and I didn't want to block this PR on it. I'd rather take an iterative approach of gradually moving more options to NodeGroupAutoscalingOptions in separate PRs (though I don't want to make any promises on when I will be able to actually implement it for ScaleDownDelay flags; I'm also happy to just review if someone else feels like taking a shot at it).

---
Details:

ScaleDownDelayAfter* are currently implemented as [a global check](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/static_autoscaler.go#L496) that result in completely skipping scale-down logic] and not as a per-node check like the 4 values in the new config. Having them per-nodegroup would require completely re-implementing them and figuring out some non-obvious questions about semantics: how should per-nodegroup ScaleDownDelayAfterDelete work? It obviously blocks one nodegroup from scaling-down, but should it be triggered only when node from that nodegroup is deleted or when any node in cluster is deleted? Same goes for ScaleDownDelayAfterAdd. It seems to me that depending on use-case you may want either behavior (block on any node - if your pods can schedule on multiple nodegroups and you just want CA to pick the best one, only block on that nodegroup if you have a dedicated nodegroup per workload and each pod can generally only go to 1 nodegroup).

Kafei59 replied:
OK I see, it's rather clear to me that CA should be first updated with "simple" modification first.
As for `ScaleDownDelayAfter*`, I understand the impact and the amount of source code edit behind, and the semantics questions. I may have a "dumb" solution but : what about adding a `ScaleDownAfterBehaviour`, with `global` effect and `local` or something ?
Anyway, thanks for the hard work, I really appreciate the effort to add some per node-group config so quickly, it will really boost our autoscaling configuration 🚀

MaciekPytel replied:
I don't think it's a dumb solution. I was considering a pretty much equivalent solution - keep the current scale-down-delay-after-* flags and add scale-down-delay-after-node-group-add/delete/etc (the name could use some work). 
This gives maximum flexibility at the cost of adding even more complexity for the user. I guess it may not be a problem if you have a hosted solution, but configuring CA is already pretty hard partly due to a huge number of flags we have. Still it may be the best solution. 

Whatever solution we decide on I'd rather do it in a separate PR though. I think this one is complex enough as is.

Kafei59 replied:
> Whatever solution we decide on I'd rather do it in a separate PR though. I think this one is complex enough as is.

And I agree :)

towca commented on cluster-autoscaler/core/scale_down_test.go:
```diff
@@ -266,6 +270,117 @@ func TestFindUnneededGPUNodes(t *testing.T) {
+			ng2 := provider.GetNodeGroup("n2").(*testprovider.TestNodeGroup)
+			ng2.SetOptions(tc.n2opts)
+
+			autoscalererr = sd.UpdateUnneededNodes(allNodes, scaleDownCandidates, time.Now(), nil)
```
nit: autoscalerErr

MaciekPytel commented on cluster-autoscaler/core/scale_down.go:
```diff
@@ -419,7 +424,23 @@ func (sd *ScaleDown) checkNodeUtilization(timestamp time.Time, node *apiv1.Node,
+	if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
+		// We should never get here as non-autoscaled nodes should not be included in scaleDownCandidates list
+		// (and the default PreFilteringScaleDownNodeProcessor would indeed filter them out).
+		klog.V(4).Infof("Skipped %s from delete considered - the node is not autoscaled", node.Name)
```
done

MaciekPytel commented on cluster-autoscaler/config/autoscaling_options.go:
```diff
@@ -45,6 +44,15 @@ type AutoscalingOptions struct {
+type AutoscalingOptions struct {
+	// NodeGroupAutoscalingOptions are default values for per NodeGroup options.
+	// They will be used any time a specific value is not provided for a given NodeGroup.
+	NodeGroupAutoscalingOptions
```
done

MaciekPytel commented on cluster-autoscaler/core/scale_down.go:
```diff
@@ -612,17 +633,24 @@ func (sd *ScaleDown) UpdateUnneededNodes(
-		if utilInfo.Utilization >= sd.context.ScaleDownGpuUtilizationThreshold {
-			return false
+		threshold, err = sd.processors.NodeGroupConfigProcessor.GetScaleDownGpuUtilizationThreshold(sd.context, nodeGroup)
+		if err != nil {
```
I'm not convinced. It would technically be 2 lines shorter, but I don't think it helps readability - I like passing through errors just after the call and not distancing it in code (admittedly here the distance would be negligible, but I like the principle).

towca replied:
Ack

MaciekPytel commented on cluster-autoscaler/processors/nodegroupconfig/node_group_config_processor_test.go:
```diff
@@ -0,0 +1,194 @@
+			want          Want
+			wantError     error
+		}{
+			"NodeGroup.GetOptions not implemented": {
```
I don't think we do this elsewhere in codebase, so my first instinct is to go for consistency. But I don't feel very strongly and I can change it if you insist.

MaciekPytel commented on cluster-autoscaler/processors/nodegroupconfig/node_group_config_processor_test.go:
```diff
@@ -0,0 +1,194 @@
+				testUnneededTime(t, p, c, ng, w, we)
+				testGpuThreshold(t, p, c, ng, w, we)
+			},
+			globalOpts: config.NodeGroupAutoscalingOptions{
```
done

MaciekPytel commented:
@towca Addressed your comment. I put the changes in a separate commit, since the fixes apply to different commits and amending them all would get tricky.

towca commented:
Thanks for addressing the comments! LGTM, but could you take a look at the second part of the remaining comment? If it's ok feel free to remove the hold.

/lgtm
/hold

MaciekPytel commented on cluster-autoscaler/core/scale_down.go:
```diff
@@ -384,6 +387,8 @@ func NewScaleDown(context *context.AutoscalingContext, clusterStateRegistry *clu
 func (sd *ScaleDown) CleanUp(timestamp time.Time) {
+	// Use default ScaleDownUnneededTime as in this context the value
+	// doesn't apply to any specific NodeGroup.
 	sd.usageTracker.CleanUp(timestamp.Add(-sd.context.ScaleDownUnneededTime))
```
You made me read usage tracker again (just when I *almost* managed to forget the horrors that lie within...). :(

I don't see any reason why it should be minimum. When we actually scale-down node we remove it explicitly from tracker, we don't need CleanUp to do it for us. If anything, setting this value to be too small will wipe usage information from tracker effectively preventing it from doing anything.

That being said - it doesn't make any sense to me that this is tied to ScaleDownUnneededTime at all. After going through the code I think we record all usage in tracker each time we run UpdateUnneededNodes (usageTracker is updated in FindNodesToRemove whether using hint or not and we call FindNodesToRemove for both currentCandidates and when seeking additional candidates). The threshold for CleanUp should really be set in a way that we remove all records that were not created in the last UpdateUnneededNodes call. This is tricky as it's related to loop time (the actual loop time and not --scan-interval) adjusted for whether we triggered scale-up or not (we don't call UpdateUnneededNodes in a loop when we trigger scale-up). 
I don't think any threshold can achieve this. I think a better way to do this would be to use some sort of logical clock that would be incremented every time we call UpdateUnneededNodes. However, this is a bigger change that isn't completely unrelated to this PR. I'm leaning towards leaving it as-is:
* It maintains the current behavior.
* There is no good value to use. Based on reading code and educated guesses (I haven't actually run any tests):
  1. Small value (1m) in a large cluster seems to have realistic risk of skipping tracker logic completely by wiping its records clean every loop (loop time >1m seems pretty realistic to me). This introduces a small risk of overzealous scale-up with rapid scale-down config.
  2. Large value means that when we actually scale-down we purge unnededNodes list based on tracker data from old simulations that are completely irrelevant. This can slow down the scale-down.
  3. I'm against using minimum because of (i) seems like the wrong direction to fail in cluster with mixed configs. I think a case could be made for both using default and using max across node groups. With (i) seeming much less likely to actually trigger than (ii) I'm leaning towards using the default.

towca replied:
Thanks for the detailed explanation, leaving as-is SGTM!

MaciekPytel replied:
It just occurred to me that I may have been missing the obvious here. Why can't we just use 0? Based on quickly looking at the code:
1. We built all relevant information in tracker in UpdateUnneededNodes.
2. We use this information to reset timer for some candidates when performing scale-down.
3. TryToScaleDown is called after running UpdateUnneededNodes. We never call TryToScaleDown without calling UpdateUnneededNodes first.

If those 3 points above are correct - why exactly do we need to keep anything in tracker between loops? Why not just fully reset it at the start of scale-down? Am I missing something? Looking for a sanity check.

BigDarkClown replied:
When I looked for the scalability issue in scale-down I also noticed that. Basically IMO the timestamps stored by UsageTracker have *no real effect* on the mechanism of usageTracker - all the relevant timestamps are stored in ScaleDown.unremovableNodes either way. The usageTracker is also rebuild in each scale-down loop, so I think that your intuition is correct.

towca commented:
/lgtm
/unhold

towca commented:
/lgtm
